File: frontend/lib/api/types/debug.ts
import type { DebugRequestStatus } from "./common";
import type { PromptContextPack } from "./prompts";

export type DebugTokenUsage = {
  today: number;
  last_7_days: number;
  last_30_days: number;
  total: number;
  unknown_usage_requests: number;
};

export type DebugReadableMessage = {
  role: string;
  content: string;
};

export type DebugReadableView = {
  system_messages: DebugReadableMessage[];
  user_messages: DebugReadableMessage[];
  request_options: Record<string, unknown>;
  context_pack: PromptContextPack | null;
  context_budget: Record<string, unknown> | null;
  context_materials: Array<Record<string, unknown>>;
  raw_content: string | null;
  parsed_payload: Record<string, unknown> | null;
  schema_error: string | null;
  embedding_summary: Record<string, unknown>;
};

export type DebugRequestLog = {
  id: string;
  project_id: string | null;
  model_kind: "llm" | "embedding";
  request_type: string;
  api_config_id: string | null;
  provider: string;
  model: string;
  status: DebugRequestStatus;
  created_at: string;
  request_body: Record<string, unknown>;
  response_body: Record<string, unknown>;
  debug_readable: DebugReadableView;
  error_message: string | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  total_tokens: number | null;
  duration_ms: number | null;
};

export type DebugSnapshot = {
  token_usage: DebugTokenUsage;
  request_logs: DebugRequestLog[];
};

export type ModelQueueItem = {
  id: string;
  kind: "llm" | "embedding";
  label: string;
  status: "queued" | "running";
  priority: "manual" | "batch" | "auto";
  model: string | null;
  queued_at: string;
  started_at: string | null;
};

export type ModelQueueSnapshot = {
  worker_count: number;
  queued_count: number;
  running_count: number;
  items: ModelQueueItem[];
};

File: backend/app/Schemas/debug.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, Field

from app.Schemas.common import DebugRequestStatus, ModelRequestKind
from app.Schemas.prompt_context import PromptContextPack


class DebugTokenUsage(BaseModel):
  today: int = 0
  last_7_days: int = 0
  last_30_days: int = 0
  total: int = 0
  unknown_usage_requests: int = 0


class DebugReadableMessage(BaseModel):
  role: str
  content: str


class DebugReadableView(BaseModel):
  system_messages: list[DebugReadableMessage] = Field(default_factory=list)
  user_messages: list[DebugReadableMessage] = Field(default_factory=list)
  request_options: dict[str, Any] = Field(default_factory=dict)
  context_pack: PromptContextPack | None = None
  context_budget: dict[str, Any] | None = None
  context_materials: list[dict[str, Any]] = Field(default_factory=list)
  raw_content: str | None = None
  parsed_payload: dict[str, Any] | None = None
  schema_error: str | None = None
  embedding_summary: dict[str, Any] = Field(default_factory=dict)


class DebugRequestLog(BaseModel):
  id: str
  project_id: str | None = None
  model_kind: ModelRequestKind = "llm"
  request_type: str
  api_config_id: str | None = None
  provider: str
  model: str
  status: DebugRequestStatus
  created_at: str
  request_body: dict[str, Any] = Field(default_factory=dict)
  response_body: dict[str, Any] = Field(default_factory=dict)
  debug_readable: DebugReadableView = Field(default_factory=DebugReadableView)
  error_message: str | None = None
  prompt_tokens: int | None = None
  completion_tokens: int | None = None
  total_tokens: int | None = None
  duration_ms: int | None = None


class DebugSnapshot(BaseModel):
  token_usage: DebugTokenUsage = Field(default_factory=DebugTokenUsage)
  request_logs: list[DebugRequestLog] = Field(default_factory=list)
